modules/core/src/main/java/org/apache/ignite/internal/util/StripedVersatileReadWriteLock.java

```diff
@@ -0,0 +1,240 @@
+     * Tries to acquire the read lock. No spinwait is used if the lock cannot be acquired immediately.
+     *
+     * <p>In contrast with async locking, invocations of this method MUST be invoked in the same thread in which the corresponding
+     * {@link #readUnlock()} will be invoked.
+     *
+     * @return {@code true} if acquired, {@code false} if write lock is already held by someone else (or someone is waiting to acquire
+     *     the write lock).
+     */
+    public boolean tryReadLock() {
+        return currentLock().tryReadLock();
+    }
+
+    /**
+     * Releases the read lock.
+     *
+     * <p>In contrast with async locking, invocations of this method MUST be invoked in the same thread in which the corresponding
+     * {@link #readLock()} is invoked.
+     *
+     * @throws IllegalMonitorStateException thrown if the read lock is not acquired by anyone.
+     */
+    public void readUnlock() {
+        currentLock().readUnlock();
+    }
+
+    private void readUnlock(int idx) {
+        locks[idx].readUnlock();
+    }
+
+    /**
+     * Acquires the write lock waiting, if needed. The thread will block until all other read and write locks are released.
+     */
+    public void writeLock() {
+        int i = 0;
+
+        // Locks must be acquired in order to avoid deadlocks.
+        for (; i < locks.length; i++) {
+            locks[i].writeLock();
+        }
+    }
+
+    private void writeUnlock0(int fromIndex) {
+        for (int i = fromIndex; i >= 0; i--) {
+            locks[i].writeUnlock();
+        }
+    }
+
+    /**
+     * Tries to acquire the write lock. Never blocks: if any lock has already been acquired by someone else, returns {@code false}
+     * immediately.
+     *
+     * @return {@code true} if the write lock has been acquired, {@code false} otherwise
+     */
+    public boolean tryWriteLock() {
+        int i = 0;
+
+        try {
+            for (; i < locks.length; i++) {
+                if (!locks[i].tryWriteLock()) {
+                    break;
+                }
+            }
+        } finally {
+            if (0 < i && i < locks.length) {
+                writeUnlock0(i - 1);
+            }
+        }
+
+        return i == locks.length;
+    }
+
+    /**
+     * Releases the write lock.
+     *
+     * @throws IllegalMonitorStateException thrown if the write lock is not acquired.
+     */
+    public void writeUnlock() {
+        writeUnlock0(locks.length - 1);
+    }
+
+    /**
+     * Executes the provided asynchronous action under protection of a read lock: that is, it first obtains a read lock
+     * asynchronously, then executes the action, and then releases the lock.
+     *
+     * @param action Action to execute.
+     * @return Action result.
+     */
+    public <T> CompletableFuture<T> inReadLockAsync(Supplier<? extends CompletableFuture<T>> action) {
+        return readLockAsync()
+                .thenCompose(index -> {
+                    return nullCompletedFuture()
```

Comment:
I'm not a fan of this approach, maybe we can compute the `index` at the beginning of this method and pass it both to `readLockAsync` and `readUnlock`? This will make this method looks similar to `inWriteLockAsync`